relayapi/internal/perform.go

```diff
@@ -0,0 +1,136 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+
+	"github.com/matrix-org/dendrite/federationapi/storage/shared"
+	"github.com/matrix-org/dendrite/internal"
+	"github.com/matrix-org/dendrite/relayapi/api"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/sirupsen/logrus"
+)
+
+// PerformRelayServerSync implements api.FederationInternalAPI
+func (r *RelayInternalAPI) PerformRelayServerSync(
+	ctx context.Context,
+	request *api.PerformRelayServerSyncRequest,
+	response *api.PerformRelayServerSyncResponse,
+) error {
+	prevEntry := gomatrixserverlib.RelayEntry{EntryID: -1}
+	asyncResponse, err := r.fedClient.GetAsyncEvents(ctx, request.UserID, prevEntry, request.RelayServer)
+	if err != nil {
+		logrus.Errorf("GetAsyncEvents: %s", err.Error())
+		return err
+	}
+	r.processTransaction(&asyncResponse.Txn)
+
+	for asyncResponse.EntriesQueued {
+		logrus.Infof("Retrieving next entry from relay, previous: %v", prevEntry)
+		asyncResponse, err = r.fedClient.GetAsyncEvents(ctx, request.UserID, prevEntry, request.RelayServer)
+		prevEntry = gomatrixserverlib.RelayEntry{EntryID: asyncResponse.EntryID}
+		if err != nil {
+			logrus.Errorf("GetAsyncEvents: %s", err.Error())
+			return err
+		}
+		r.processTransaction(&asyncResponse.Txn)
+	}
+
+	return nil
+}
+
+// PerformStoreAsync implements api.RelayInternalAPI
+func (r *RelayInternalAPI) PerformStoreAsync(
+	ctx context.Context,
+	request *api.PerformStoreAsyncRequest,
+	response *api.PerformStoreAsyncResponse,
+) error {
+	logrus.Warnf("Storing transaction for %v", request.UserID)
+	receipt, err := r.db.StoreAsyncTransaction(ctx, request.Txn)
+	if err != nil {
+		logrus.Errorf("db.StoreAsyncTransaction: %s", err.Error())
+		return err
+	}
+	err = r.db.AssociateAsyncTransactionWithDestinations(
+		ctx,
+		map[gomatrixserverlib.UserID]struct{}{
+			request.UserID: {},
+		},
+		request.Txn.TransactionID,
+		receipt)
+
+	return err
+}
+
+// QueryAsyncTransactions implements api.RelayInternalAPI
+func (r *RelayInternalAPI) QueryAsyncTransactions(
+	ctx context.Context,
+	request *api.QueryAsyncTransactionsRequest,
+	response *api.QueryAsyncTransactionsResponse,
+) error {
+	logrus.Infof("QueryAsyncTransactions for %s", request.UserID.Raw())
+	if request.PreviousEntry.EntryID >= 0 {
+		logrus.Infof("Cleaning previous entry (%v) from db for %s",
+			request.PreviousEntry.EntryID,
+			request.UserID.Raw(),
+		)
+		prevReceipt := shared.NewReceipt(request.PreviousEntry.EntryID)
+		err := r.db.CleanAsyncTransactions(ctx, request.UserID, []*shared.Receipt{&prevReceipt})
+		if err != nil {
+			logrus.Errorf("db.CleanAsyncTransactions: %s", err.Error())
+			return err
```

Comment:
It would be because of a database issue so it's probably not worth carrying on.
Also it guarantees that previous entry was removed before requesting the next transaction. Otherwise, we would ignore the failure and return the exact same transaction again.